File: Client/wwwroot/js/pageJsInvokeService.js
window._registerPage = function (type, page) {
  window._pagesToInvoke ??= {};

  window._pagesToInvoke[type] = page;
};

window._unregisterPage = function (type) {
  if (window._pagesToInvoke) {
    ref = window._pagesToInvoke[type];
    delete window._pagesToInvoke[type];

    return ref;
  }
};

window.getPageRef = function (type) {
  if (window._pagesToInvoke) {
    return window._pagesToInvoke[type];
  }
};

window._unregisterAll = function () {
  if (window._pagesToInvoke) {
    refs = Object.values(window._pagesToInvoke);
    window._pagesToInvoke = {};

    return refs;
  } else {
    return [];
  }
};

File: Client/Services/PageJsInvokeServices/PageJsInvokeService.cs
using Client.Pages;
using Microsoft.JSInterop;
using System.Reflection;

namespace Client.Services
{
	public class PageJsInvokeService : IPageJsInvokeService, IDisposable
	{
		private readonly IJSRuntime _jsRuntime;
		private bool _inited = false;
		private bool _disposed = false;

		public PageJsInvokeService(IJSRuntime jsRuntime)
		{
			_jsRuntime = jsRuntime;
		}

		public async Task RegisterAsync<T>(T page) where T : class
		{
			if (!_inited)
			{
				await _jsRuntime.InvokeVoidAsync("import", "./js/pageJsInvokeService.js");
				_inited = true;
			}

			DotNetObjectReference<T>? pageRef = DotNetObjectReference.Create(page);
			await _jsRuntime.InvokeAsync<string>("_registerPage", typeof(T).Name, pageRef);
		}

		public async Task UnregisterAsync<T>() where T : class
		{
			if (!_inited)
			{
				await _jsRuntime.InvokeVoidAsync("import", "./js/pageJsInvokeService.js");
				_inited = true;
			}

			var pageRef = await _jsRuntime.InvokeAsync<DotNetObjectReference<T>?>("_unregisterPage", typeof(T).Name);
			pageRef?.Dispose();
		}

		public async void Dispose()
		{
			if (_disposed)
				return;

			if (_inited)
			{
				var pageRefs = await _jsRuntime.InvokeAsync<IEnumerable<IDisposable>>("_unregisterAll");

				foreach (var pageRef in pageRefs)
				{
					pageRef.Dispose();
				}
			}

			_disposed = true;
			GC.SuppressFinalize(this);
		}
	}
}